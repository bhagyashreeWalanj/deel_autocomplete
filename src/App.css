.App-header {
  background-color: #a6e2ff;
  min-height: 100vh;
  /* display: flex;
  flex-direction: column; */
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
  padding-top: 4rem;
}


.container{
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  width: auto;
  /* margin-top: 4rem; */
}
.header{
  color: chocolate;
  font-size: 26px;
  font-family: Arial, Helvetica, sans-serif;
}
.autocomplete{
  width: 500px;
  border: 1px solid #d4d4d4;
  border-radius: 10px;
  background: #fff;
}

.autocomplete input{
  width: 27rem;
  border: none;
  margin: 10px 0;
  font-size: 16px;
  padding: 10px 10px 0 10px;

}

input:focus{
  outline: none;
}

.suggestions{
  list-style: none;
  max-height: 30rem;
  /* height:500px; */
  overflow: auto;
  overflow-y:scroll;
  color: #282c34;
  margin: 0;
  padding: 0;
  border-top: 1px solid #d4d4d4;
  background-color: #f7f7f7;
  border-radius: 0 0 10px 10px;
  scroll-behavior: smooth; /* nice smooth movement */
  font-size: 16px;

}

.suggestions li{
  list-style: none;
  padding: 10px;
  cursor: pointer;
}
strong {
  background: #29c076;
  color: #f7f7f7;
  font-weight: inherit;
}
.suggestions li:focus{
  background:LIGHTSALMON;
  outline:none;
}
.autocomplete li:hover{
  background-color: lightskyblue;
}



.option {
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 500px;
  padding: 5px;
}

.closeButton{
  width: 2rem;
  height: 1.5rem;
  margin: 0.8rem 0.3rem 0 0;
  float: right;
  cursor: pointer;
  color: tomato;
}