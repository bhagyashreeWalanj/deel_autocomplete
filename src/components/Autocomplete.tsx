import React, { useState, useEffect, useRef } from 'react'
import { TiLocation, TiDelete } from 'react-icons/ti'
import { ISuggestion } from '../interface/ISuggestion'

export interface IAutocomplete {
  countries: ISuggestion[]
}

const Autocomplete = ({ countries }: IAutocomplete) => {
  const [searchValue, setSearchValue] = useState<string>('')
  const [newArray, setNewArray] = useState<ISuggestion[]>([])
  const [showSuggestion, setShowSuggestion] = useState(false)

  const autoCompleteRef = useRef<HTMLDivElement>(null)

  let filteredCountries: ISuggestion[] = []
  if (newArray.length !== 0) {
    filteredCountries = newArray
  } else {
    countries.forEach((country) => {
      if (country.option.toLowerCase().includes(searchValue.toLowerCase())) {
        filteredCountries.push({
          option: country.option,
          highlighted: country.highlighted,
        })
      }
    })
  }

  const handleOnChangeText = (e: any) => {
    let searchText: string = e.target.value
    searchText.length <= 0 ? setShowSuggestion(false) : setShowSuggestion(true)

    setSearchValue(searchText)
    let finalArray: ISuggestion[] = []

    countries.forEach((option) => {
      const element: string = option.option
      let matchStartIndex: number = 0
      if (
        (matchStartIndex = element
          .toLowerCase()
          .indexOf(searchText.toLowerCase())) > -1
      ) {
        let matchStart = element.slice(0, matchStartIndex)
        let highlighted = element.substring(
          matchStartIndex,
          matchStartIndex + searchText.length,
        )
        let matchEnd: string = element.substring(
          matchStartIndex + searchText.length,
        )

        if (matchStartIndex >= 0) {
          let matchText = `${matchStart}<strong>${highlighted}</strong>${matchEnd}`

          finalArray.push({
            option: element,
            highlighted: matchText,
          })
        }
      }
    })
    setNewArray(finalArray)
  }

  // handle onClick of list element to country select from the list and set on input
  const handleSuggestion = (suggestion: string) => {
    setSearchValue(suggestion)
    setShowSuggestion(false)
  }

  // To hide the list if mouse click on screen
  const handleOutsideClick = (event: any) => {
    const { current: wrap } = autoCompleteRef
    if (wrap && !wrap.contains(event.target)) {
      setShowSuggestion(false)
    }
  }

  // delete the text from input
  const handleDelete = () => {
    setSearchValue('')
    setShowSuggestion(true)
  }

  useEffect(() => {
    document.addEventListener('mousedown', handleOutsideClick)
    return () => {
      document.removeEventListener('mousedown', handleOutsideClick)
    }
  }, [])

  return (
    <div className="autocomplete" ref={autoCompleteRef}>
      <input
        type="text"
        id="autocomplete_input"
        value={searchValue}
        onChange={(e) => handleOnChangeText(e)}
        placeholder="Search Country..."
        autoComplete="off"
      />

      {searchValue !== '' ? (
        <TiDelete className="closeButton" onClick={handleDelete} />
      ) : (
        ''
      )}

      {showSuggestion ? (
        filteredCountries.length === 0 ? (
          <ul className="suggestions">
            <li>No Data Found</li>
          </ul>
        ) : (
          <ul className="suggestions">
            {filteredCountries.map((country, index) => {
              return (
                <li
                  className="options"
                  id={`#suggest_${index}`}
                  onClick={() => handleSuggestion(country.option)}
                  key={country.option}
                >
                  <TiLocation style={{ marginRight: '0.8rem' }} />
                  <span
                    dangerouslySetInnerHTML={{ __html: country.highlighted }}
                  ></span>
                </li>
              )
            })}
          </ul>
        )
      ) : (
        ''
      )}
    </div>
  )
}

export default Autocomplete
