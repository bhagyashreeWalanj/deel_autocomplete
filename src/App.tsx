import React, { useState, useEffect } from 'react'
import './App.css'
import Autocomplete from './components/Autocomplete'
import { ISuggestion } from './interface/ISuggestion'
import { API_URL } from './API/api'

function App() {
  const [countries, setCountries] = useState<ISuggestion[]>([])
  const fetchCountries = async () => {
    const response = await fetch(API_URL, { method: 'GET' })
    const responseInJson = await response.json()
    return responseInJson
  }

  useEffect(() => {
    fetchCountries()
      .then((response) => {
        let jsonArr: ISuggestion[] = []
        response.data.map((item: any) => {
          jsonArr.push({
            option: item.name,
            highlighted: item.name,
          })
          return item.name
        })
        setCountries(jsonArr)
      })
      .catch((err) => console.log('ERROR : ', err))
  }, [])
  return (
    <div className="App-header">
      <div className="container">
        <h1 className="header">REACT AUTOCOMPLETE</h1>
        <p style={{ color: 'black' }}>Type 'a' for display list</p>
        <Autocomplete countries={countries} />
      </div>
    </div>
  )
}

export default App
