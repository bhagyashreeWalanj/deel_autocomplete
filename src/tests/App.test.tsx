import App from '../App'
import Autocomplete from '../components/Autocomplete'
import { shallow } from 'enzyme'

describe('render all components', () => {
  const app = shallow(<App />)

  it('should match the snapshot', () => {
    expect(app).toMatchSnapshot()
  })
  it('App Component contains Autocomplete component', () => {
    expect(app.contains(<Autocomplete countries={[]} />)).toBe(true)
  })
})
