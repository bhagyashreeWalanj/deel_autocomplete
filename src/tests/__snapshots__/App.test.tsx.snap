// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render all components should match the snapshot 1`] = `
<div
  className="App-header"
>
  <div
    className="container"
  >
    <h1
      className="header"
    >
      REACT AUTOCOMPLETE
    </h1>
    <p
      style={
        Object {
          "color": "black",
        }
      }
    >
      Type 'a' for display list
    </p>
    <Autocomplete
      countries={Array []}
    />
  </div>
</div>
`;
